### PR TITLE
Public method for creating a Session

### LINKED ISSUE #124: Make Session function public for external calls
Currently, I have implemented a Session object myself. Can these two methods be opened for external self control calls.

Make `new_process_mux` or `new_native_mux` public .

```diff

mpl Session {
    #[cfg(feature = "process-mux")]
+    pub(super) fn new_process_mux(tempdir: TempDir) -> Self {
        Self(SessionImp::ProcessImpl(process_impl::Session::new(tempdir)))
    }

    #[cfg(feature = "native-mux")]
+    pub(super) fn new_native_mux(tempdir: TempDir) -> Self {
        Self(SessionImp::NativeMuxImpl(native_mux_impl::Session::new(
            tempdir,
        )))
    }
}
```

[About code link ref:](https://github.com/openssh-rust/openssh/blob/3319a8162a/src/session.rs#L56-L65)

### PR DESCRIPTION
**Pull request by baoyachi**
fix #124

### CODE REVIEW

**Comment by jonhoo** [comment]
This change is [Reviewable](https://reviewable.io/reviews/openssh-rust/openssh/128)

File: src/session.rs [hunk 1]
```diff
@@ -53,13 +53,107 @@ pub struct Session(SessionImp);
...
+    /// #[derive(Clone, Debug, Default)]
+    /// pub struct CustomSessionBuilder {
+    ///    user: Option<String>,
+    ///    port: Option<String>,
+    ///    password: Option<String>,
+    /// }
+    ///
+    /// let builder = CustomSessionBuilder::default().password("123456");
+    /// let (builder, destination) = builder.resolve("ssh://jon@ssh.thesquareplanet.com:222");
+    /// let tempdir = builder.launch_master(destination).await?;
```
Comment (NobodyXu) [inline src/session.rs:81]: I would rather the examples to be much simpler and pass the compilation.

File: src/session.rs [hunk 2]
```diff
@@ -53,13 +53,107 @@ pub struct Session(SessionImp);
...
 
 impl Session {
+    /// The method for creating a [`Session`] and externally control the creation of TempDir.
+    ///
+    /// By using the built-in [`SessionBuilder`] in openssh, or a custom SessionBuilder,
+    /// create a TempDir.
+    ///
+    /// # Examples
+    ///
+    /// ```ignore
```
Comment (NobodyXu) [inline src/session.rs:63]: I'm ok with `no_run` but not ok with ignoring the example altogether:

```suggestion
    /// ```no_run
```

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: src/session.rs (hunk 1 above)
Comment (baoyachi) [reply src/session.rs:81]: That's a good idea.

File: src/session.rs (hunk 2 above)
Comment (baoyachi) [reply src/session.rs:63]: I will write an example to modify it

File: src/builder.rs [hunk 3]
```diff
@@ -286,7 +286,7 @@ impl SessionBuilder {
         Ok(f(tempdir))
     }
 
-    fn resolve<'a, 'b>(&'a self, mut destination: &'b str) -> (Cow<'a, Self>, &'b str) {
+    pub fn resolve<'a, 'b>(&'a self, mut destination: &'b str) -> (Cow<'a, Self>, &'b str) {
```
Comment (NobodyXu) [inline src/builder.rs:289]: You need to provide doc for this public method.

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: src/builder.rs [hunk 4]
```diff
@@ -324,7 +324,7 @@ impl SessionBuilder {
         (Cow::Owned(with_overrides), destination)
     }
 
-    async fn launch_master(&self, destination: &str) -> Result<TempDir, Error> {
+    pub async fn launch_master(&self, destination: &str) -> Result<TempDir, Error> {
```
Comment (NobodyXu) [inline src/builder.rs:327]: Same for this, doc is required.

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: src/builder.rs [hunk 5]
```diff
@@ -286,7 +286,22 @@ impl SessionBuilder {
...
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use std::borrow::Cow;
+    /// use openssh::SessionBuilder;
+    /// let b = SessionBuilder::default();
+    /// let (b, d) = b.resolve("ssh://test-user@127.0.0.1:2222");
+    /// assert_eq!(d, "127.0.0.1");
+    /// assert_eq!(b, Cow::Borrowed(SessionBuilder::default()
```
Comment (NobodyXu) [inline src/builder.rs:300]: You could instead, add a getter for retrieving `port` and `user` of `SessioBuilder`.

Comment (NobodyXu) [inline src/builder.rs:300]: You could instead, add a getter for retrieving `port` and `user` of `SessioBuilder`.

File: src/builder.rs [hunk 6]
```diff
@@ -55,7 +55,7 @@ fn clean_history_control_dir(socketdir: &Path, prefix: &str) -> io::Result<()> {
 }
 
 /// Build a [`Session`] with options.
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, PartialEq)]
```
Comment (NobodyXu) [inline src/builder.rs:58]: I would not like `PartialEq` to be added to `SessionBuilder`, it should be an opaque struct and we do not provide any guarantee on how it is stored or the order of them.

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: src/builder.rs (hunk 5 above)
Comment (baoyachi) [reply src/builder.rs:300]: port or use not pub use in doc

Comment (NobodyXu) [reply src/builder.rs:300]: I mean adding a getter for that:

```
/// Return the user set in builder.
pub fn get_user(&self) -> Option<&str> {
    self.user.as_deref()
}

Comment (NobodyXu) [reply src/builder.rs:300]: I mean adding a getter for that:

```
/// Return the user set in builder.
pub fn get_user(&self) -> Option<&str> {
    self.user.as_deref()
}

File: src/builder.rs [hunk 7]
```diff
@@ -286,7 +296,21 @@ impl SessionBuilder {
...
     }
 
-    fn resolve<'a, 'b>(&'a self, mut destination: &'b str) -> (Cow<'a, Self>, &'b str) {
+    /// [`SessionBuilder`] support for `destination` parsing.
+    /// The format of `destination` is the same as the `destination` argument to `ssh`.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use std::borrow::Cow;
```
Comment (NobodyXu) [inline src/builder.rs:305]:
```suggestion
```

File: src/builder.rs [hunk 8]
```diff
@@ -324,7 +348,8 @@ impl SessionBuilder {
         (Cow::Owned(with_overrides), destination)
     }
 
-    async fn launch_master(&self, destination: &str) -> Result<TempDir, Error> {
+    /// Create ssh master session and return [`TempDir`]
```
Comment (NobodyXu) [inline src/builder.rs:351]:
```suggestion
    /// Create ssh master session and return [`TempDir`] which
    /// contains the ssh control socket.
```

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: src/builder.rs (hunk 7 above)
Comment (baoyachi) [reply src/builder.rs:305]: @NobodyXu  .How to find quick this code warning ?

**Review by NobodyXu — APPROVED** [review]

**Pull request merged**